src/rgw/rgw_common.cc

```diff
@@ -1118,26 +1118,35 @@ struct perm_state_from_req_state : public perm_state_base {
 };
 
 Effect eval_or_pass(const DoutPrefixProvider* dpp,
-		    const boost::optional<Policy>& policy,
-		    const rgw::IAM::Environment& env,
-		    boost::optional<const rgw::auth::Identity&> id,
-		    const uint64_t op,
-		    const ARN& resource,
-				boost::optional<rgw::IAM::PolicyPrincipal&> princ_type=boost::none) {
+                    const boost::optional<Policy>& policy,
+                    const rgw::IAM::Environment& env,
+                    boost::optional<const rgw::auth::Identity&> id,
+                    const uint64_t op,
+                    const ARN& resource,
+                    const rgw_owner& bucket_owner,
+                    bool restrict_public_buckets,
+                    boost::optional<rgw::IAM::PolicyPrincipal&> princ_type=boost::none) {
   if (!policy)
     return Effect::Pass;
-  else
-    return policy->eval(env, id, op, resource, princ_type);
+
+  // If RestrictPublicBuckets is enabled and the policy allows public access, deny the request if the requester is not in the bucket owner account
+  if (restrict_public_buckets && rgw::IAM::is_public(*policy) && !id->is_owner_of(bucket_owner)) {
+    return Effect::Deny;
```

Comment:
adding this to the common function `eval_or_pass()` means it applies to every kind of policy, including identity and session policies

looking at [PutPublicAccessBlock](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutPublicAccessBlock.html#API_PutPublicAccessBlock_RequestSyntax), BlockPublicPolicy and RestrictPublicBuckets both specify _bucket_ policy:
> Specifies whether Amazon S3 should restrict public bucket policies for this bucket. Setting this element to TRUE restricts access to this bucket to only AWS service principals and authorized users within this account if the bucket has a public policy.

do you see any other docs that talk about identity/session policies?

to make this specific to bucket policy, i think you can just add checks for `bucket_policy` in `verify_bucket_permission()` and `verify_object_permission()`

Reply:
right, I guess you should not be able to create any identity/session policies with Principal `*` at all. so `rgw::IAM::is_public(*policy)` should never be true?

Reply:
I have moved this out to `evaluate_iam_policies()` in respect to only being considered by `resource_policy` anyway as you said, that doesn't look like the right place to be called.
I'm unsure about `verify_topic_permission()` and whether we should consider it or not.
What do you think, should still get it exported further to `verify_bucket_permission()` and `verify_object_permission()` explicitly?

Reply:
i'm not sure whether we're strict about Principal in those policies, but yeah